src/main/java/com/google/sps/servlets/GroupsServlet.java

```diff
@@ -26,74 +30,76 @@
-      // Not displaying the group ID, so remove it.
-      group.remove(GROUP_ID_PROPERTY);
-
-      // Turn the Map values into Strings rather than Objects for the soy template to read
-      ImmutableList<Map<String, String>> groups = ImmutableList.of((Map) group);
-      
-      // Each group has its own map which points to its info and all maps are passed into the template as a list
-      // This will make it easier when groups are queried from Datastore
-      ImmutableMap<String, ImmutableList<Map<String, String>>> data = ImmutableMap.of(GROUPS_KEY, groups);
-
-      String htmlString = SoyRendererUtils.getOutputString(GROUPS_SOY_FILE, GROUPS_TEMPLATE_NAMESPACE, data);
-
-      response.setContentType(CONTENT_TYPE_HTML);
-      response.getWriter().println(htmlString);
-    } catch (Exception entityError) {
-       ServletUtils.printBadRequestError(response, ENTITY_ERROR_MESSAGE);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    Query groupsQuery = new Query(GROUP_KIND);
+    PreparedQuery preparedGroupsQuery = datastore.prepare(groupsQuery); 
+
+    ImmutableList.Builder<ImmutableMap<String, String>> groupsListBuilder = new ImmutableList.Builder<>();
+    for (Entity group : preparedGroupsQuery.asIterable()) {
+      ImmutableMap<String, String> groupMap = ImmutableMap.of(
+          GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
+          GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY),
+          GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY));
+      groupsListBuilder.add(groupMap);
     }
+
+    ImmutableList<ImmutableMap<String, String>> groupsList = groupsListBuilder.build();
+
+    // Each group has its own map which points to its info and all maps are passed into the template as a list
+    // This will make it easier when groups are queried from Datastore
+    ImmutableMap<String, ImmutableList<ImmutableMap<String, String>>> groupsData = 
+        ImmutableMap.of(GROUPS_KEY, groupsList);
+
+    String groupsPageHtml = SoyRendererUtils.getOutputString(GROUPS_SOY_FILE, GROUPS_TEMPLATE_NAMESPACE, groupsData);
+
+    response.setContentType(CONTENT_TYPE_HTML);
+    response.getWriter().println(groupsPageHtml);
   }
 
   /**
    * Create a group. Parameters are name, description (optional), and image.
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    //TODO: Unhardcode. Only one hardcoded group will be created and displayed temporarily, but 
-    // will change to enable more groups.
-    String groupId = createGroup(GROUP_NAME, GROUP_IMAGE, GROUP_DESCRIPTION);
+    String name = getParameter(request, GROUP_NAME_PROPERTY);
+    String image = getParameter(request, GROUP_IMAGE_PROPERTY);
 
-    response.setContentType(CONTENT_TYPE_HTML);
+    if (Strings.isNullOrEmpty(name) || Strings.isNullOrEmpty(image)){
```

Comment:
hopefully you shouldn't need this because I set all fields to be required on the form so the form cannot be submitted until the user has typed something for every input. I don't know if it'd be good practice to keep this extra layer of security in there or not?

Reply:
doing it just in case someone tries to do a POST request and create a group from outside the frontend, so not completely necessary.